provider/k8s/k8s.go

```diff
@@ -31,23 +36,120 @@ type K8s struct {
 	ctx context.Context
 }
 
+type K8sClient struct {
+	// The k8s provider used when we work with the manifest files.
+	k8sProvider *K8s
+	// DeploymentFiles files provided from the cli.
+	DeploymentFiles []string
+	// Vaiables to subtitude in the DeploymentFiles.
+	// These are also used when the command requires some variables that are not provided by the deployment file.
+	DeploymentVars map[string]string
+	// DeploymentFile content after substituting the variables filename is used as the map key.
+	deploymentsContent []provider.ResourceFile
+
+	ctx context.Context
+}
+
 // New returns a k8s client that can apply and delete resources.
-func New(ctx context.Context, config clientcmdapi.Config) (*K8s, error) {
-	restConfig, err := clientcmd.NewDefaultClientConfig(config, &clientcmd.ConfigOverrides{}).ClientConfig()
+func New(ctx context.Context, clientset *kubernetes.Clientset) *K8s {
+	return &K8s{
+		ctx: ctx,
+		clt: clientset,
+	}
+}
+
+// NewK8sClient returns a k8s client that can apply and delete resources.
+func NewK8sClient() (*K8sClient, error) {
+	config, err := rest.InClusterConfig()
 	if err != nil {
-		return nil, errors.Wrapf(err, "config error")
+		return nil, errors.Wrapf(err, "k8s config error")
 	}
-	clientset, err := kubernetes.NewForConfig(restConfig)
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
 	if err != nil {
-		return nil, errors.Wrapf(err, "client error")
+		return nil, errors.Wrapf(err, "k8s client error")
 	}
 
-	return &K8s{
-		ctx: ctx,
-		clt: clientset,
+	ctx := context.Background()
+	return &K8sClient{
+		k8sProvider:    New(ctx, clientset),
+		DeploymentVars: make(map[string]string),
+		ctx:            ctx,
 	}, nil
 }
 
+// DeploymentsParse parses the deployment files and saves the result as bytes grouped by the filename.
+// Any variables passed to the cli will be replaced in the resources files following the golang text template format.
+func (c *K8sClient) DeploymentsParse(*kingpin.ParseContext) error {
```

Comment:
DeploymentsParse and applyTemplateVars is too much code to be duplicated in gke and k8s and eventually it will diverge, we need to put it somewhere we can reuse. Maybe provider/provider.go is a good place for it.

Reply:
I have moved them to provider.go in https://github.com/prometheus/prombench/pull/139/commits/a7855f3d960e2db0d54589d29b727f50960b3e81